src/Nest/XPack/MachineLearning/EstimateModelMemory/EstimateModelMemoryRequest.cs

```diff
@@ -0,0 +1,127 @@
+	{
+		IAnalysisConfig IEstimateModelMemoryRequest.AnalysisConfig { get; set; }
+		IOverallCardinality IEstimateModelMemoryRequest.OverallCardinality { get; set; }
+		IMaxBucketCardinality IEstimateModelMemoryRequest.MaxBucketCardinality { get; set; }
+
+		/// <inheritdoc />
+		public EstimateModelMemoryDescriptor<TDocument> AnalysisConfig(Func<AnalysisConfigDescriptor<TDocument>, IAnalysisConfig> selector) =>
+			Assign(selector, (a, v) => a.AnalysisConfig = v?.Invoke(new AnalysisConfigDescriptor<TDocument>()));
+
+		/// <inheritdoc />
+		public EstimateModelMemoryDescriptor<TDocument> OverallCardinality(Func<OverallCardinalityDescriptor<TDocument>, IPromise<IOverallCardinality>> analyzerSelector) =>
+			Assign(analyzerSelector, (a, v) => a.OverallCardinality = v?.Invoke(new OverallCardinalityDescriptor<TDocument>())?.Value);
+
+		/// <inheritdoc />
+		public EstimateModelMemoryDescriptor<TDocument> MaxBucketCardinality(Func<MaxBucketCardinalityDescriptor<TDocument>, IPromise<IMaxBucketCardinality>> analyzerSelector) =>
+			Assign(analyzerSelector, (a, v) => a.MaxBucketCardinality = v?.Invoke(new MaxBucketCardinalityDescriptor<TDocument>())?.Value);
+	}
+
+	[JsonFormatter(typeof(VerbatimDictionaryKeysFormatter<OverallCardinality, IOverallCardinality, Field, long>))]
+	public interface IOverallCardinality : IIsADictionary<Field, long> { }
+
+	public class OverallCardinality : IsADictionaryBase<Field, long>, IOverallCardinality
+	{
+		public OverallCardinality() { }
+
+		public OverallCardinality(IDictionary<Field, long> container) : base(container) { }
+
+		public OverallCardinality(Dictionary<Field, long> container) : base(container) { }
+
+		public void Add(Field field, long cardinality) => BackingDictionary.Add(field, cardinality);
+	}
+
+	public class OverallCardinality<T> : OverallCardinality where T : class
+	{
+		public void Add<TValue>(Expression<Func<T, TValue>> field, long cardinality) => BackingDictionary.Add(field, cardinality);
+	}
+
+	public class OverallCardinalityDescriptor<T> : IsADictionaryDescriptorBase<OverallCardinalityDescriptor<T>, IOverallCardinality, Field, long> where T : class
+	{
+		public OverallCardinalityDescriptor() : base(new OverallCardinality()) { }
+
+		public OverallCardinalityDescriptor<T> Field(Field field, long cardinality) => Assign(field, cardinality);
+
+		public OverallCardinalityDescriptor<T> Field<TValue>(Expression<Func<T, TValue>> field, long cardinality) => Assign(field, cardinality);
+	}
+
+	[JsonFormatter(typeof(VerbatimDictionaryKeysFormatter<MaxBucketCardinality, IMaxBucketCardinality, Field, long>))]
+	public interface IMaxBucketCardinality : IIsADictionary<Field, long> { }
+
+	public class MaxBucketCardinality : IsADictionaryBase<Field, long>, IMaxBucketCardinality
+	{
+		public MaxBucketCardinality() { }
+
+		public MaxBucketCardinality(IDictionary<Field, long> container) : base(container) { }
+
+		public MaxBucketCardinality(Dictionary<Field, long> container) : base(container) { }
```

Comment:
Does it need both interface and concrete ctor overloads?

Reply:
Good Q, will remove the concrete